packages/ember-data/lib/system/record-array-manager.js

```diff
@@ -121,11 +122,28 @@ export default Ember.Object.extend({
   */
   updateFilterRecordArray: function(array, filter, typeClass, record) {
     var shouldBeInArray = filter(record.getRecord());
-    var recordArrays = this.recordArraysForRecord(record);
+    var key = guidFor(record);
+    if (shouldBeInArray && shouldBeInArray.then) {
+      array.lastFilterPromises[key] = shouldBeInArray;
+      shouldBeInArray.then((result) => {
+        // fighting race conditions here
+        if (shouldBeInArray !== array.lastFilterPromises[key]) {
```

Comment:
@pangratz hmm.. seems like your are right
@igorT something like this needs to be added:
```javascript
shouldBeInArray.catch( => {
  // fighting race conditions here
  if (shouldBeInArray !== array.lastFilterPromises[key]) {
    return;
  }

  delete array.lastFilterPromises[key];
  this._updateFilterRecordArray(array, false, typeClass, record);
});

```